### PR TITLE
Run on Julia 1.5 (by building sysimg) and 1.7 (with LBT) in the same codebase

### PR DESCRIPTION
**Pull request by ViralBShah**
Improved version of #62 

Use LBT for Julia releases that have it, and the sysimg method otherwise. Julia 1.6 is not yet compatible with with the MKL sysimage building.

### CODE REVIEW

**Comment by ViralBShah** [comment]
The MKL.jl package now works now for 1.5 and 1.7 (1.6 is broken, unrelated to this PR). Once reviewed, we should be able to merge it, making it easier for more people to try out MKL.

@KristofferC @staticfloat @andreasnoack @dmbates

**Comment by chriselrod** [comment]
On 1.7:
```julia
julia> using LinearAlgebra

julia> BLAS.vendor()
:openblas64

julia> using MKL
ERROR: InitError: UndefVarError: is_available not defined
Stacktrace:
 [1] getproperty
   @ ./Base.jl:26 [inlined]
 [2] __init__()
   @ MKL ~/.julia/dev/MKL/src/MKL.jl:39
 [3] _include_from_serialized(path::String, depmods::Vector{Any})
   @ Base ./loading.jl:674
 [4] _require_search_from_serialized(pkg::Base.PkgId, sourcepath::String)
   @ Base ./loading.jl:760
 [5] _require(pkg::Base.PkgId)
   @ Base ./loading.jl:998
 [6] require(uuidkey::Base.PkgId)
   @ Base ./loading.jl:914
 [7] require(into::Module, mod::Symbol)
   @ Base ./loading.jl:901
during initialization of module MKL

julia> versioninfo()
Julia Version 1.7.0-DEV.624
Commit 7367dcd4f4* (2021-02-27 14:38 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, tigerlake)
Environment:
  JULIA_NUM_THREADS = 8
```

**Comment by giordano** [comment]
An out-of-date version of `MKL_jll`?

**Comment by chriselrod** [comment]
That was with the master branch.

**Comment by giordano** [comment]
Default branch is `main`, not `master`

File: Project.toml [hunk 1]
```diff
@@ -12,7 +13,6 @@ PackageCompiler = "1"
 julia = "1.3"
```
Comment (giordano) [inline Project.toml:13]: Should probably add compat for MKL to avoid confusion
```julia
MKL = "2021"
```

**Comment by chriselrod** [comment]
```julia
julia> using MKL

julia> BLAS.vendor()
:openblas64

julia> BLAS.set_num_threads(4);

signal (11): Segmentation fault
in expression starting at REPL[9]:1
MKL_SET_NUM_THREADS at /home/chriselrod/.julia/artifacts/a8e009985328801a84c9af6610f94f77a7c12852/lib/libmkl_intel_ilp64.so.1 (unknown line)
lbt_set_num_threads at /workspace/srcdir/libblastrampoline/src/threading.c:86
lbt_set_num_threads at /home/chriselrod/Documents/languages/julia/usr/share/julia/stdlib/v1.7/LinearAlgebra/src/lbt.jl:116 [inlined]
set_num_threads at /home/chriselrod/Documents/languages/julia/usr/share/julia/stdlib/v1.7/LinearAlgebra/src/blas.jl:124
unknown function (ip: 0x7f4bcab1fe42)
```

**Comment by ViralBShah** [comment]
We should just update to the latest MKL version here.

**Comment by ViralBShah** [comment]
The `vendor()` API should hopefully be possible to remove by the time we get to the 1.7 release.

In the meanwhile, see: https://github.com/JuliaLang/julia/blob/f2b5b6e3d45fb12369336814b6be08a035c07b7c/stdlib/LinearAlgebra/src/blas.jl#L91

**Comment by ViralBShah** [comment]
@chriselrod Can you post a separate issue for the set_num_threads crash?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -47,4 +47,5 @@ jobs:
           using Pkg
           Pkg.develop(PackageSpec(path="."))
           Pkg.build("MKL"; verbose=true)
+	  Pkg.test("MKL")
```
Comment (giordano) [inline .github/workflows/ci.yml:50]: This is run by the following step

**Comment by ViralBShah** [comment]
@giordano Any idea what that windows failure is about? Not an error I have seen before and doesn't seem to be related to this package.

**Comment by giordano** [comment]
No, I don't know the error, I'd need to debug interactively with tmate to understand what's going on, but I'd note that the path `C:\D:\a\MKL.jl\MKL.jl` looks suspiciously wrong

**Comment by ViralBShah** [comment]
Maybe @KristofferC might know what is going on with Windows.

**Comment by crstnbr** [comment]
As per @ViralBShah's request on Slack I tried MKL#lbt2 in combination with julia#master on macOS:
```Julia
julia> versioninfo()
Julia Version 1.7.0-DEV.625
Commit f2b5b6e3d4 (2021-02-27 20:54 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin19.6.0)
  CPU: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, skylake)
Environment:
  JULIA_EDITOR = subl
  JULIA_NUM_THREADS = 6
  JULIA_PKG_SERVER = eu-central.pkg.julialang.org
```

Basically I see the same behaviour as described by @chriselrod. `BLAS.vendor()` is wrong, `BLAS.set_num_threads` segfaults after `using MKL` but apart from that things seem to work just fine!

```Julia
julia> using LinearAlgebra

julia> BLAS.get_config()
LinearAlgebra.BLAS.LBTConfig(LinearAlgebra.BLAS.LBTLibraryInfo[LinearAlgebra.BLAS.LBTLibraryInfo("/Applications/Julia-1.7.app/Contents/Resources/julia/bin/../lib/julia/libopenblas64_.dylib", Ptr{Nothing} @0x00007f9fa7498ce0, "64_", UInt8[0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff  …  0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x01], :ilp64, :plain)], [:f2c_capable], ["LAPACKE_c_nancheck", "LAPACKE_cbbcsd", "LAPACKE_cbbcsd_work", "LAPACKE_cbdsqr", "LAPACKE_cbdsqr_work", "LAPACKE_cgb_nancheck", "LAPACKE_cgb_trans", "LAPACKE_cgbbrd", "LAPACKE_cgbbrd_work", "LAPACKE_cgbcon"  …  "zunmlq_", "zunmql_", "zunmqr_", "zunmr2_", "zunmr3_", "zunmrq_", "zunmrz_", "zunmtr_", "zupgtr_", "zupmtr_"])

julia> using BenchmarkTools

julia> M = K = N = 200; A = rand(M,K); B = rand(K,N); C = Matrix{Float64}(undef, M, N);

julia> @benchmark mul!($C, $A, $B)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     80.471 μs (0.00% GC)
  median time:      90.480 μs (0.00% GC)
  mean time:        94.923 μs (0.00% GC)
  maximum time:     352.737 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> using MKL

julia> BLAS.get_config()
LinearAlgebra.BLAS.LBTConfig(LinearAlgebra.BLAS.LBTLibraryInfo[LinearAlgebra.BLAS.LBTLibraryInfo("/Users/crstnbr/.julia/artifacts/073ff95e2c63501547247d6e1321bf4ee2a78933/lib/libmkl_rt.1.dylib", Ptr{Nothing} @0x00007f9fa97fb4b0, "", UInt8[0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff  …  0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x01], :ilp64, :plain)], [:f2c_capable], ["LAPACKE_c_nancheck", "LAPACKE_cbbcsd", "LAPACKE_cbbcsd_work", "LAPACKE_cbdsqr", "LAPACKE_cbdsqr_work", "LAPACKE_cgb_nancheck", "LAPACKE_cgb_trans", "LAPACKE_cgbbrd", "LAPACKE_cgbbrd_work", "LAPACKE_cgbcon"  …  "zunmlq_", "zunmql_", "zunmqr_", "zunmr2_", "zunmr3_", "zunmrq_", "zunmrz_", "zunmtr_", "zupgtr_", "zupmtr_"])

julia> @benchmark mul!($C, $A, $B)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     61.837 μs (0.00% GC)
  median time:      71.509 μs (0.00% GC)
  mean time:        74.482 μs (0.00% GC)
  maximum time:     208.506 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> BLAS.set_num_threads(4)

signal (11): Segmentation fault: 11
in expression starting at REPL[10]:1
MKL_SET_NUM_THREADS at /Users/crstnbr/.julia/artifacts/073ff95e2c63501547247d6e1321bf4ee2a78933/lib/libmkl_intel_ilp64.1.dylib (unknown line)
Allocations: 9614540 (Pool: 9610950; Big: 3590); GC: 27
[1]    52490 segmentation fault  julia-latest --project=.
```

**Comment by crstnbr** [comment]
BTW, what's the strategy if a users wants to always use MKL over OpenBLAS? Should he put `using MKL` in his `startup.jl`?

**Comment by ViralBShah** [comment]
Regarding `BLAS.vendor()`, see my comment in https://github.com/JuliaLinearAlgebra/MKL.jl/pull/63#issuecomment-787217528. Check `BLAS.get_config()` to get the details of which BLAS is loaded. It gives you a lot more information than `vendor()` did. The threading issue has been opened - will get fixed shortly.

For now, yes, add `using MKL` to the startup.jl if you always want to use MKL. @staticfloat mentioned to me that we expect to get Preferences.jl into 1.7 (possibly), and when that happens, we'll use it to set a BLAS preference.

File: src/MKL.jl [hunk 3]
```diff
@@ -16,41 +16,39 @@ end
...
-
+function mklnorm(x::Vector{Float64})
+    ccall((:dnrm2_, libmkl_rt), Float64,
+          (Ref{MKLBlasInt}, Ptr{Float64}, Ref{MKLBlasInt}),
+          length(x), x, 1)
+end
 
-# Include installation routines as well
-include("install.jl")
+VERSION > v"1.7.0-DEV.623" && include("install.jl")
```
Comment (KristofferC) [inline src/MKL.jl:52]: Isn't this the wrong direction? We want to `include` the install file if we do not have access to LBT.

Comment (ViralBShah) [reply src/MKL.jl:52]: That is what the code does.

https://github.com/JuliaLinearAlgebra/MKL.jl/blob/259493b6e811c60fc2436b8efdd65665a91c5bd7/src/MKL.jl#L54

Comment (KristofferC) [reply src/MKL.jl:52]: This is on 1.6:

```
julia> JULIA_VER_NEEDED = v"1.7.0-DEV.641"
v"1.7.0-DEV.641"

julia> VERSION > JULIA_VER_NEEDED
false
```

so the LHS is false and the `include` is not executed on 1.6. And on 1.6 I cannot execute this:

```
julia> VERSION
v"1.6.2"

julia> using MKL

julia> MKL.enable_mkl_startup()
ERROR: UndefVarError: enable_mkl_startup not defined
```

However, on 1.7 it is defined:

```
julia> MKL.enable_mkl_startup()
[ Info: Splicing in code to load MKL in /Users/kristoffercarlsson/julia/usr/bin/../share/julia/base/sysimg.jl
ERROR: UndefVarError: USEBLAS64 not defined
```

So I think the conditional should be reversed here.

**Pull request merged**